scipy/ndimage/src/nd_image.c

```diff
@@ -189,54 +81,79 @@ NA_NewArray(void *buffer, enum NPY_TYPES type, int ndim, npy_intp *shape)
     return result;
 }
 
-/* Convert an input array of any type, not necessarily contiguous */
+/* Converts a Python array-like object into a behaved input array. */
 static int
 NI_ObjectToInputArray(PyObject *object, PyArrayObject **array)
 {
-    *array = NA_InputArray(object, NPY_NOTYPE, NPY_ARRAY_ALIGNED|NPY_ARRAY_NOTSWAPPED);
-    return *array ? 1 : 0;
+    int flags = NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED;
+    *array = (PyArrayObject *)PyArray_CheckFromAny(object, NULL, 0, 0, flags,
+                                                   NULL);
+    return *array != NULL;
 }
 
-/* Convert an input array of any type, not necessarily contiguous */
+/* Like NI_ObjectToInputArray, but with special handling for Py_None. */
 static int
 NI_ObjectToOptionalInputArray(PyObject *object, PyArrayObject **array)
 {
     if (object == Py_None) {
         *array = NULL;
         return 1;
-    } else {
-        *array = NA_InputArray(object, NPY_NOTYPE, NPY_ARRAY_ALIGNED|NPY_ARRAY_NOTSWAPPED);
-        return *array ? 1 : 0;
     }
+    return NI_ObjectToInputArray(object, array);
 }
 
-/* Convert an output array of any type, not necessarily contiguous */
+/* Converts a Python array-like object into a behaved output array. */
 static int
 NI_ObjectToOutputArray(PyObject *object, PyArrayObject **array)
 {
-    *array = NA_OutputArray(object, NPY_NOTYPE, NPY_ARRAY_ALIGNED|NPY_ARRAY_NOTSWAPPED);
-    return *array ? 1 : 0;
+    int flags = NPY_ARRAY_BEHAVED_NS | NPY_ARRAY_UPDATEIFCOPY;
+    /*
+     * This would also be caught by the PyArray_CheckFromAny call, but
+     * we check it explicitly here to provide a saner error message.
+     */
+    if (PyArray_Check(object) &&
+            !PyArray_ISWRITEABLE((PyArrayObject *)object)) {
+        PyErr_SetString(PyExc_ValueError, "output array is read-only.");
+    return 0;
+    }
+    /*
+     * If the input array is not aligned or is byteswapped, this call
+     * will create a new aligned, native byte order array, and copy the
+     * contents of object into it. For an output array, the copy is
+     * unnecessary, so this could be optimized. It is very easy to not
+     * do NPY_ARRAY_UPDATEIFCOPY right, so we let NumPy do it for us
+     * and pay the performance price.
+     */
+    *array = (PyArrayObject *)PyArray_CheckFromAny(object, NULL, 0, 0, flags,
+                                                   NULL);
+    return *array != NULL;
 }
 
-/* Convert an output array of any type, not necessarily contiguous */
+/* Like NI_ObjectToOutputArray, but with special handling for Py_None. */
 static int
 NI_ObjectToOptionalOutputArray(PyObject *object, PyArrayObject **array)
 {
     if (object == Py_None) {
         *array = NULL;
         return 1;
-    } else {
-        *array = NA_OutputArray(object, NPY_NOTYPE, NPY_ARRAY_ALIGNED|NPY_ARRAY_NOTSWAPPED);
-        return *array ? 1 : 0;
     }
+    return NI_ObjectToOutputArray(object ,array);
```

Comment:
This line was missing the `return` before the function call, so we where in "undefined behavior" territory...